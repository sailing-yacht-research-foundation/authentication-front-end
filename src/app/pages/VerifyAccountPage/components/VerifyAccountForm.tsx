import * as React from 'react';
import { Input, Form } from 'antd';
import { Auth } from 'aws-amplify';
import { useHistory } from 'react-router';
import { toast } from 'react-toastify';
import { SyrfFormButton } from 'app/components/SyrfForm';

export const VerifyAccountForm = () => {
    const history = useHistory<any>();

    React.useEffect(() => {
        let code = new URLSearchParams(history.location.search).get("code");
        let email = new URLSearchParams(history.location.search).get("email");

        if (!history?.location?.state?.state?.email && !(code && email)) {
            history?.push('/not-found')
        } else if (code && email) {
            verifyAccount(email, code);
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const onFinish = (values) => {
        const { code } = values;
        const email = history?.location?.state?.state?.email;

        verifyAccount(email, code);
    }

    const showVerifySuccessAndRedirectToLogin = () => {
        history.push('/signin');
        toast.success('Account verified, please login!');
    }

    const verifyAccount = (email, code) => {
        Auth.confirmSignUp(email, code)
            .then(response => {
                showVerifySuccessAndRedirectToLogin();
            }).catch(error => {
                if (error.code !== 'NotAuthorizedException') // don't show when user already confirmed and only want to re-verify email.
                    toast.error(error.message);
            });

        Auth.currentAuthenticatedUser().then(user => { // this case is for re-verify email.
            Auth.verifyCurrentUserAttributeSubmit('email', code).then(() => {
                showVerifySuccessAndRedirectToLogin();
                Auth.signOut();
            }).catch(error => {
                toast.error(error.message);
            });
        }).catch(e => {

        });
    }

    const resendConfirmationCode = () => {
        const email = history?.location?.state?.state?.email;

        Auth.resendSignUp(email).then(response => {
            toast.success('Confirmation code sent!');
        }).catch(error => {
            toast.error(error.message);
        })
    }

    return (
        <Form
            layout={'vertical'}
            name="basic"
            initialValues={{ remember: true }}
            onFinish={onFinish}
        >
            <Form.Item
                label="Verification code"
                name="code"
                rules={[{ required: true }]}
            >
                <Input />
            </Form.Item>

            <Form.Item>
                <SyrfFormButton type="primary" htmlType="submit">
                    Verify my account
                </SyrfFormButton>
                <div style={{ marginTop: '10px', textAlign: 'right' }}>
                    <span> Could not receive the code? &nbsp; <a style={{ float: 'right' }} href="/" onClick={(e) => {
                        e.preventDefault();
                        resendConfirmationCode();
                    }}>resend</a></span>
                </div>
            </Form.Item>
        </Form>
    );
}
