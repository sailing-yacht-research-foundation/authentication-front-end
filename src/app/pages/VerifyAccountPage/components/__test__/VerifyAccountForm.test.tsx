import * as React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { VerifyAccountForm } from '../VerifyAccountForm';

const shallowRenderer = createRenderer();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: () => ({
    pathname: "localhost:3002/"
  })
}));

describe('<VerifyAccountForm />', () => {
  it('should render and matches the snaphshot', () => {
    shallowRenderer.render(<VerifyAccountForm />);
    const renderedOutput = shallowRenderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});