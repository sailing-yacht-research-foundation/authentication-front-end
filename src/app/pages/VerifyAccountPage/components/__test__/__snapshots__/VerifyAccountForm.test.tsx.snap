// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerifyAccountForm /> should render and matches the snaphshot 1`] = `
<ForwardRef(InternalForm)
  initialValues={
    Object {
      "remember": true,
    }
  }
  layout="vertical"
  name="basic"
  onFinish={[Function]}
>
  <FormItem
    label="Verification code"
    name="code"
    rules={
      Array [
        Object {
          "required": true,
        },
      ]
    }
  >
    <Input
      type="text"
    />
  </FormItem>
  <FormItem>
    <Styled(Button)
      htmlType="submit"
      type="primary"
    >
      Verify my account
    </Styled(Button)>
    <div
      style={
        Object {
          "marginTop": "10px",
          "textAlign": "right",
        }
      }
    >
      <span>
         Could not receive the code? Â  
        <a
          href="/"
          onClick={[Function]}
          style={
            Object {
              "float": "right",
            }
          }
        >
          resend
        </a>
      </span>
    </div>
  </FormItem>
</ForwardRef(InternalForm)>
`;
